### ISSUE TITLE
Update two models

### ISSUE THREAD

**Issue by wangjiawen2013** [tracel-ai/burn#2630]
Hi,
I meet a use case. I constuct two models (model1 and model2). The input of model1 is the output of model2. Loss_1 is calculate from model1's forward and another independent tensor(input_y). Loss_2 is calculated only from model1's forward. I want to update model1 using loss_1 and then update model2 using loss_2 . This is my code:
```rust
type B = Autodiff<NdArray>;
let mut optimizer_1 = RmsPropConfig::new().init::<B, Model1<B>>();
let mut optimizer_2 = RmsPropConfig::new().init::<B, Model2<B>>();

// update model1, input and input_y are  tensors that independent from both model2 and model1
let input_x = model2.foward(input.clone()).detach();

let loss_1 = model1.forward(input_x).mean() - model1.forward(input_y).mean();
let grads = loss_1.backward();
let grads = GradientsParams::from_grads(grads, &model1);
model1 = optimizer_1.step(config.lr, model1, grads);

// update model2
let input_x = model2.foward(input);

let loss_2 = -model1.forward(input_x).mean();
let grads = loss_2.backward();
let grads = GradientsParams::from_grads(grads, &model2);
model2 = optimizer_2.step(config.lr, model2, grads);
```
Though I can run the code, I am not sure whether it works internally as expected . I don't know where to use detach to divide the two updating process. Perhaps the two updating process will interact with each other, which is I want to avoid. Who can kindly tell is the logic right ?

**Comment by laggui** [comment]
To prevent gradients from flowing back into `model2` when computing `loss_1`, just detach the input to your `model1.forward(...)`

```rust
let input_x = model2.forward(input);

// Prevent gradients from flowing back into model2 when computing loss_1,
// ensuring that only model1 is updated during the first backward pass.
let loss_1 = model1.forward(input_x.detach()).mean() - model1.forward(input_y).mean();

let grads = loss_1.backward();
let grads = GradientsParams::from_grads(grads, &model1);
model1 = optimizer_d.step(config.lr, model1, grads);
```

Now, to update `model2` using `loss_2`, I'm not sure because what you describe and the code in the snippet don't match.
```rust
// model1 is involved in computing the loss_2 here, is this a mistake?
let loss_2 = -model1.forward(input_x).mean();

let grads = loss_2.backward();
let grads = GradientsParams::from_grads(grads, &model2);
model2 = optimizer_d.step(config.lr, model2, grads);
```

**Comment by wangjiawen2013** [comment]
Thanks!
I fixed a typo, `optimizer_d` should be `optimizer_1` for model1 and `optimizer_2` for model2. There is no other mistakes in the code. Model1 is involved in computing the loss_2 because `let loss_2 = -model1.forward(input_x).mean()` and `let input_x = model2.forward(input)`. I think it essentially equals to `let loss_2 = -model1.forward(model2.forward(input)).mean()`, more specifically, both model 1 and model2 are involved in computing loss_2.  In this step, I want to update only the parameters of model2 but not model1. And this is why initialized two optimizers (`optimizer_1` and `optimizer_2`), because I want to update only model1 and then update only model2 without changing the parameter of model1.

**Comment by wangjiawen2013** [comment]
In summary, My 
```
let input_x = model2.forward(input);

// Only update model1's parameters
let loss_1 = model1.forward(input_x.clone().detach()).mean() - model1.forward(input_y).mean();
......
// Only update models'2 parameters
let loss_2 = -model1.forward(input_x).mean()
......

**Comment by wangjiawen2013** [comment]
Perhaps the `optimizer_1` and `optimizer_2` will interact with each other, or `GradientsParams::from_grads(grads, &model1)` 
 and `GradientsParams::from_grads(grads, &model2)` will interact with each other, which can cause the mis-updating of the two models.

**Comment by laggui** [comment]
Ahh ok, with the updated snippet following my comment this should work as intended I believe!

- `model2.forward(input).detach()` ensures that gradients from `loss_1` do no propagate back into `model2` during the backward pass and `optimizer2.step()`
- even if `loss_2` computes gradients for all parameters involved, using `GradientsParams::from_grads(grads, &model2)` will filter the gradients for `model2` and then `optimizer2.step(lr, model2, grads)` will only update `model2`'s parameters
